### PR TITLE
fix(nuxt): skip hydration mismatches with client components

### PR DESCRIPTION
**Pull request by huang-julien**
### 🔗 Linked issue

https://nuxt.com/docs/guide/directory-structure/components#paired-with-a-client-component

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Hi 👋 this PR solves the hydration issues between `.client` and `.server` components.

This allow to have different html on initial loads, even if both does not have the same number of root nodes.

The "workaround" is to render a `staticVNode` before hydration and then render the `client` component when mounted.

Directives and attributes are removed before hydration (and set back on mounted) since `staticVNodes` canno't inherit attributes or receive directives. 

About the component `template` property, I didn't found a working solution for the runtime compiler, but since this isn't supported yet, I think this maybe removed or left as it is or even implemented with the support vue full build PR. 

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/app/components/client-only.mjs [hunk 1]
```diff
@@ -73,3 +94,50 @@ export function createClientOnly (component) {
...
 }
+
+function getStaticVNode (vnode) {
+  const fragment = getFragmentHTML(vnode.el)
+
+  if (fragment.length === 0) {
+    return null
+  }
+  return createStaticVNode(fragment.join(''), fragment.length)
+}
```
Comment (danielroe) [inline packages/nuxt/src/app/components/client-only.mjs:105]: Is there any way you think we can cut down on the extra code here and reduce bundle size when using client components? Is any of this shared with our server-component implementation?

Comment (huang-julien) [reply packages/nuxt/src/app/components/client-only.mjs:105]: sure we can :+1: ! Server components are just a copy/paste of theses lines of code 😃

File: packages/nuxt/src/app/components/client-only.ts [hunk 2]
```diff
@@ -51,8 +53,23 @@ export function createClientOnly<T extends ComponentOptions> (component: T) {
...
   clone.setup = (props, ctx) => {
+    const instance = getCurrentInstance()!
+
+    // @ts-expect-error internal
+    const inheritAttrs = instance.inheritAttrs
+    // remove exsting directives during hydration
+    const directives = extractDirectives(instance)
+    // prevent attrs inheritance since a staticVNode is rendered before hydration
+    // @ts-expect-error internal
+    instance.inheritAttrs = false
```
Comment (huang-julien) [inline packages/nuxt/src/app/components/client-only.ts:64]: not sure if it's a good idea to rely on internals

Comment (danielroe) [reply packages/nuxt/src/app/components/client-only.ts:64]: yes, I'm not sure I would like to do that either. 🤔

**Review by antfu — APPROVED** [review]
Smart :) LGTM. Now I feel at some point this should be a built-in component of Vue.

**Review by danielroe — APPROVED** [review]

**Pull request merged**